### ISSUE TITLE
Nil pointer panic in sdk. 

### ISSUE THREAD

**Issue by rshade** [pulumi/pulumi#11727]
We are getting the following error, and it is panic'ing pulumi cli. 
```
         [stdout]          Diagnostics:
         [stdout]            pulumi:pulumi:Stack (<REDACTED>):
         [stdout]              panic: runtime error: invalid memory address or nil pointer dereference
         [stdout]              [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xb6cd78]
         [stdout]              goroutine 1798 [running]:
         [stdout]              github.com/pulumi/pulumi/sdk/v3/go/pulumi.mergeDependencies({0x0, 0x1, 0xc000570000?}, {0xc000791a10, 0x1, 0x40c0ad?})
         [stdout]              	/home/nonroot/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.39.0/go/pulumi/types.go:212 +0x218
         [stdout]              github.com/pulumi/pulumi/sdk/v3/go/pulumi.(*OutputState).fulfillValue(0xc0005d60e0, {0x230f020?, 0xc000791a00?, 0x0?}, 0x1, 0x0, {0xc000791a10, 0x1, 0x1}, {0x0, ...})
         [stdout]              	/home/nonroot/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.39.0/go/pulumi/types.go:198 +0x39e
         [stdout]              github.com/pulumi/pulumi/sdk/v3/go/pulumi.(*OutputState).ApplyTWithContext.func1()
         [stdout]              	/home/nonroot/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.39.0/go/pulumi/types.go:518 +0x61b
         [stdout]              created by github.com/pulumi/pulumi/sdk/v3/go/pulumi.(*OutputState).ApplyTWithContext
         [stdout]              	/home/nonroot/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.39.0/go/pulumi/types.go:495 +0x411
         [stdout]              exit status 2
         [stdout]           
         [stdout]              error: an unhandled error occurred: waiting for RPCs: rpc error: code = Unavailable desc = error reading from server: EOF
```
We believe it has to do with the following line: 
https://github.com/pulumi/pulumi/blob/v3.39.0/sdk/go/pulumi/types.go#L212

**Comment by Zaid-Ajaj** [comment]
Hi @rshade, this looks interesting because the expression `range <slice>` will not error if `<slice>` is nil so it must be something else. Can you please provide a piece of code so we can reproduce the problem?

**Comment by squaremo** [comment]
What version of the CLI is in question?

**Comment by Frassle** [comment]
It's in the go sdk, not the cli.

**Comment by squaremo** [comment]
> It's in the go sdk, not the cli.

OK, what is the version of the SDK? The code snippet link is to a particular commit; I want to establish whether that is the code _at the version used_.

**Comment by justinvp** [comment]
The version is in the stack trace: `@v3.39.0`

**Comment by squaremo** [comment]
> The version is in the stack trace: `@v3.39.0`

So it is, I didn't notice that. I've edited the description so that it explicitly points at the code as at the version used (the line is the same -- @rshade if you deliberately used the SHA1 so it would show a preview, apologies, my change has lost that).

**Comment by t0yv0** [comment]
The Pulumi Platform team took a deep dive on the issue yesterday and I want to summarize it in a quick update.

The issue arises from a race leading to a torn slice being passed into mergeDependencies as the first argument (`len(x) == 1 && x == nil`).

Since v3.39.0 there was we had [work done](https://github.com/pulumi/pulumi/pull/11248) to pass the test suite under the race detector, which introduced locks around the code in question. It is our understanding that this issue should be fixed in version v3.46.1 and higher.

We have confidence but not 100% certainty here that the issue is fixed because we failed to construct a repro case - a solid repro would definitely be highly appreciated! To mitigate any further issues, we introduced checks that detect memory corruption earlier and produce better panic stack traces in case it happens so that it is easier to reproduce the issue. This work is https://github.com/pulumi/pulumi/pull/11776 and https://github.com/pulumi/pulumi/pull/11779
